src/cc/tests/distributed_test.cc

```diff
@@ -610,6 +610,214 @@ TEST(DistributedTest, FullNeighborsMultipleServers)
+}
+
+TEST(DistributedTest, NeighborSampleMultipleTypesNeighborsSpreadAcrossPartitions)
+{
+    auto environment =
+        CreateMultiServerSplitNeighborsEnvironment("NeighborSampleMultipleTypesNeighborsSpreadAcrossPartitions");
+    auto &c = *environment.second;
+
+    std::vector<snark::NodeId> nodes = {1};
+    std::vector<snark::Type> types = {0, 1};
+    int count = 6;
+    std::vector<snark::NodeId> neighbor_nodes(count * nodes.size(), -1);
+    std::vector<snark::Type> neighbor_types(count * nodes.size(), -1);
+    std::vector<float> neighbor_weights(count * nodes.size(), -1);
+
+    c.WeightedSampleNeighbor(13, std::span(nodes), std::span(types), count, std::span(neighbor_nodes),
+                             std::span(neighbor_types), std::span(neighbor_weights), 0, 0, 0);
+
+    EXPECT_EQ(std::vector<snark::NodeId>({4, 6, 7, 6, 5, 5}), neighbor_nodes);
+    EXPECT_EQ(std::vector<snark::Type>({0, 1, 1, 1, 1, 1}), neighbor_types);
+    EXPECT_EQ(std::vector<float>({1.f, 1.5f, 3.f, 1.5f, 1.f, 1.f}), neighbor_weights);
+}
+
+TEST(DistributedTest, UniformNeighborSampleMultipleTypesNeighborsSpreadAcrossPartitions)
+{
+    auto environment =
+        CreateMultiServerSplitNeighborsEnvironment("UniformNeighborSampleMultipleTypesNeighborsSpreadAcrossPartitions");
+    auto &c = *environment.second;
+
+    std::vector<snark::NodeId> nodes = {1};
+    std::vector<snark::Type> types = {0, 1};
+    int count = 6;
+    std::vector<snark::NodeId> neighbor_nodes(count * nodes.size(), -1);
+    std::vector<snark::Type> neighbor_types(count * nodes.size(), -1);
+    std::vector<uint64_t> total_neighbor_counts(nodes.size());
+
+    c.UniformSampleNeighbor(true, 17, std::span(nodes), std::span(types), count, std::span(neighbor_nodes),
+                            std::span(neighbor_types), 0, 2);
+    EXPECT_EQ(std::vector<snark::NodeId>({7, 4, 0, 0, 0, 0}), neighbor_nodes);
+    EXPECT_EQ(std::vector<snark::Type>({1, 0, 2, 2, 2, 2}), neighbor_types);
+}
+
+std::pair<ServerList, std::shared_ptr<snark::GRPCClient>> CreateMultiServerSplitFeaturesEnvironment(
+    std::string name, std::vector<std::vector<float>> f0, std::vector<std::vector<float>> f1,
+    std::vector<std::vector<float>> f2)
+{
+    const size_t num_servers = 2;
+    ServerList servers;
+    std::vector<std::shared_ptr<grpc::Channel>> channels;
+
+    TestGraph::MemoryGraph m1;
+    m1.m_nodes.push_back(TestGraph::Node{.m_id = 0, .m_type = 0, .m_weight = 1.0f, .m_float_features = f0});
+    m1.m_nodes.push_back(TestGraph::Node{.m_id = 1, .m_type = 1, .m_weight = 1.0f, .m_float_features = f1});
+    m1.m_nodes.push_back(TestGraph::Node{.m_id = 2, .m_type = -1, .m_weight = 1.0f});
+    TestGraph::MemoryGraph m2;
+    m2.m_nodes.push_back(TestGraph::Node{.m_id = 1, .m_type = -1});
+    m2.m_nodes.push_back(TestGraph::Node{.m_id = 2, .m_type = 2, .m_float_features = f2});
+    std::vector<TestGraph::MemoryGraph> test_graphs = {m1, m2};
```

Comment:
Please, add one more case with `m1` and `m2` swapped to verify we correctly skip partitions before a feature is found.

Reply:
Yes this is already the case, in m1, id = 1 is a real node and id = 2 is a ghost node.